stdlib/stdlib-external/char.ts

```diff
@@ -0,0 +1,87 @@
+import { throwError } from './ascutils/grainRuntime'
+import { GRAIN_ERR_MALFORMED_UTF8 } from './ascutils/errors'
+import { GRAIN_GENERIC_HEAP_TAG_TYPE } from './ascutils/tags'
+import { allocateChar } from './ascutils/dataStructures'
+
+// Algorithm from https://encoding.spec.whatwg.org/#utf-8-decoder
+export function code(c: u32): u32 {
+  c = c ^ GRAIN_GENERIC_HEAP_TAG_TYPE
+
+  let codePoint: u32 = 0
+  let bytesSeen: u32 = 0
+  let bytesNeeded: u32 = 0
+  let lowerBoundary: u8 = 0x80
+  let upperBoundary: u8 = 0xBF
+
+  let offset: u32 = 0
+
+  while (true) {
+    const byte = load<u8>(c + offset++, 4)
+    if (bytesNeeded === 0) {
+      if (byte >= 0x00 && byte <= 0x7F) {
+        return byte << 1
+      } else if (byte >= 0xC2 && byte <= 0xDF) {
+        bytesNeeded = 1
+        codePoint = byte & 0x1F
+      } else if (byte >= 0xE0 && byte <= 0xEF) {
+        if (byte === 0xE0) lowerBoundary = 0xA0
+        if (byte === 0xED) upperBoundary = 0x9F
+        bytesNeeded = 2
+        codePoint = byte & 0xF
+      } else if (byte >= 0xF0 && byte <= 0xF4) {
+        if (byte === 0xF0) lowerBoundary = 0x90
+        if (byte === 0xF4) upperBoundary = 0x8F
+        bytesNeeded = 3
+        codePoint = byte & 0x7
+      } else {
+        throwError(GRAIN_ERR_MALFORMED_UTF8, 0, 0)
+        unreachable()
+      }
+      continue
+    }
+    if (!(lowerBoundary <= byte && byte <= upperBoundary)) {
+      throwError(GRAIN_ERR_MALFORMED_UTF8, 0, 0)
+    }
+    lowerBoundary = 0x80
+    upperBoundary = 0xBF
+    codePoint = (codePoint << 6) | (byte & 0x3F)
+    bytesSeen++
+    if (bytesSeen === bytesNeeded) {
+      return codePoint << 1
+    }
+  }
+}
+
+// Algorithm from https://encoding.spec.whatwg.org/#utf-8-encoder
+export function fromCode(code: u32): u32 {
+  code = code >>> 1
+  if (code < 0x80) {
+    let char = allocateChar()
```

Comment:
Not really. This way feels a tad more clear to me since all of the code for the "fast" case is contained within the `if`. We'd get a few bytes back if we moved it up though.

Reply:
Seems fine. Was just curious.